File: xdialog/src/main/java/com/cl/xdialog/base/XController.java
package com.cl.xdialog.base;

import android.content.DialogInterface;
import android.os.Parcel;
import android.os.Parcelable;
import android.view.Gravity;
import android.view.View;

import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.cl.xdialog.R;
import com.cl.xdialog.listener.OnBindViewListener;
import com.cl.xdialog.listener.OnViewClickListener;

import java.io.Serializable;

/**
 * 项目：xDialog
 * 版权：蒲公英公司 版权所有
 * 作者：Arry
 * 版本：1.0
 * 创建日期：2019-08-17
 * 描述：
 *  数据保存封装的容器类
 * 修订历史：
 */
public class XController<A extends XBaseAdapter> implements Parcelable, Serializable {

    private FragmentManager fragmentManager;
    private int layoutRes;
    private int height;
    private int width;
    private float dimAmount;
    private int gravity;
    private String tag;
    private int[] ids;
    private boolean isCancelableOutside;
    private OnViewClickListener onViewClickListener;
    private OnBindViewListener onBindViewListener;
    private A adapter;
    private XBaseAdapter.OnAdapterItemClickListener adapterItemClickListener;
    private int orientation;
    private int dialogAnimationRes;
    private View dialogView;
    private DialogInterface.OnDismissListener onDismissListener;
    private DialogInterface.OnKeyListener onKeyListener;


    //////////////////////////////////////////Parcelable持久化//////////////////////////////////////////////////////
    public XController() {
    }

    protected XController(Parcel in) {
        layoutRes = in.readInt();
        height = in.readInt();
        width = in.readInt();
        dimAmount = in.readFloat();
        gravity = in.readInt();
        tag = in.readString();
        ids = in.createIntArray();
        isCancelableOutside = in.readByte() != 0;
        orientation = in.readInt();
    }

    public static final Creator<XController> CREATOR = new Creator<XController>() {
        @Override
        public XController createFromParcel(Parcel in) {
            return new XController(in);
        }

        @Override
        public XController[] newArray(int size) {
            return new XController[size];
        }
    };

    //内容描述接口,不用管
    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(layoutRes);
        dest.writeInt(height);
        dest.writeInt(width);
        dest.writeFloat(dimAmount);
        dest.writeInt(gravity);
        dest.writeString(tag);
        dest.writeIntArray(ids);
        dest.writeByte((byte) (isCancelableOutside ? 1 : 0));
        dest.writeInt(orientation);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////

    //get
    public FragmentManager getFragmentManager() {
        return fragmentManager;
    }

    public int getLayoutRes() {
        return layoutRes;
    }

    public void setLayoutRes(int layoutRes) {
        this.layoutRes = layoutRes;
    }

    public int getHeight() {
        return height;
    }

    public int getWidth() {
        return width;
    }

    public void setWidth(int mWidth) {
        this.width = mWidth;
    }

    public float getDimAmount() {
        return dimAmount;
    }

    public int getGravity() {
        return gravity;
    }

    public String getTag() {
        return tag;
    }

    public int[] getIds() {
        return ids;
    }

    public boolean isCancelableOutside() {
        return isCancelableOutside;
    }

    public OnViewClickListener getOnViewClickListener() {
        return onViewClickListener;
    }

    public int getOrientation() {
        return orientation;
    }

    public OnBindViewListener getOnBindViewListener() {
        return onBindViewListener;
    }

    public DialogInterface.OnDismissListener getOnDismissListener() {
        return onDismissListener;
    }

    public DialogInterface.OnKeyListener getOnKeyListener() {
        return onKeyListener;
    }

    public View getDialogView() {
        return dialogView;
    }

    //列表
    public A getAdapter() {
        return adapter;
    }

    public void setAdapter(A adapter) {
        this.adapter = adapter;
    }

    public XBaseAdapter.OnAdapterItemClickListener getAdapterItemClickListener() {
        return adapterItemClickListener;
    }

    public void setAdapterItemClickListener(XBaseAdapter.OnAdapterItemClickListener adapterItemClickListener) {
        this.adapterItemClickListener = adapterItemClickListener;
    }

    public int getDialogAnimationRes() {
        return dialogAnimationRes;
    }

    /**************************************************************************
     */
    public static class TParams<A extends XBaseAdapter> {
        public FragmentManager mFragmentManager;
        public int mLayoutRes;
        public int mWidth;
        public int mHeight;
        public float mDimAmount = 0.2f;
        public int mGravity = Gravity.CENTER;
        public String mTag = "TDialog";
        public int[] ids;
        public boolean mIsCancelableOutside = true;
        public OnViewClickListener mOnViewClickListener;
        public OnBindViewListener bindViewListener;
        public int mDialogAnimationRes = 0;//弹窗动画
        //列表
        public A adapter;
        public XBaseAdapter.OnAdapterItemClickListener adapterItemClickListener;
        public int listLayoutRes;
        public int orientation = LinearLayoutManager.VERTICAL;//默认RecyclerView的列表方向为垂直方向
        public View mDialogView;//直接使用传入进来的View,而不需要通过解析Xml
        public DialogInterface.OnDismissListener mOnDismissListener;
        public DialogInterface.OnKeyListener mKeyListener;

        public void apply(XController tController) {
            tController.fragmentManager = mFragmentManager;
            if (mLayoutRes > 0) {
                tController.layoutRes = mLayoutRes;
            }
            if (mDialogView != null) {
                tController.dialogView = mDialogView;
            }
            if (mWidth > 0) {
                tController.width = mWidth;
            }
            if (mHeight > 0) {
                tController.height = mHeight;
            }
            tController.dimAmount = mDimAmount;
            tController.gravity = mGravity;
            tController.tag = mTag;
            if (ids != null) {
                tController.ids = ids;
            }
            tController.isCancelableOutside = mIsCancelableOutside;
            tController.onViewClickListener = mOnViewClickListener;
            tController.onBindViewListener = bindViewListener;
            tController.onDismissListener = mOnDismissListener;
            tController.dialogAnimationRes = mDialogAnimationRes;
            tController.onKeyListener =mKeyListener;

            if (adapter != null) {
                tController.setAdapter(adapter);
                if (listLayoutRes <= 0) {//使用默认的布局
                    tController.setLayoutRes(R.layout.dialog_recycler);
                } else {
                    tController.setLayoutRes(listLayoutRes);
                }
                tController.orientation = orientation;
            } else {
                if (tController.getLayoutRes() <= 0 && tController.getDialogView() == null) {
                    throw new IllegalArgumentException("请先调用setLayoutRes()方法设置弹窗所需的xml布局!");
                }
            }
            if (adapterItemClickListener != null) {
                tController.setAdapterItemClickListener(adapterItemClickListener);
            }

            //如果宽高都没有设置,则默认给弹窗提供宽度为600
            if (tController.width <= 0 && tController.height <= 0) {
                tController.width = 600;
            }
        }
    }
}

File: xdialog/src/main/java/com/cl/xdialog/base/XController.kt
package com.cl.xdialog.base

import android.content.DialogInterface
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import androidx.fragment.app.FragmentManager
import com.cl.xdialog.listener.OnAdapterItemClickListener
import com.cl.xdialog.listener.OnBindViewListener
import com.cl.xdialog.listener.OnViewClickListener
import java.io.Serializable

/**
 * 项目：xDialog
 * 版权：蒲公英公司 版权所有
 * 作者：Arry
 * 版本：1.0
 * 创建日期：2019-08-17
 * 描述：
 * 数据保存封装的容器类
 * 修订历史：
 */
class XController<A : XBaseAdapter<*>?> : Parcelable, Serializable {
    ////////////////////////////////////////////////////////////////////////////////////////////////
    //get
    var fragmentManager: FragmentManager? = null
    var layoutRes = 0
    var height = 0
    var width = 0
    var dimAmount = 0f
    var gravity = 0
    var tag: String? = null
    var ids: IntArray? = null
    var isCancelableOutside = false
    var onViewClickListener: OnViewClickListener? = null
    var onBindViewListener: OnBindViewListener? = null


    //列表
    var adapter: A? = null
        private set
    var adapterItemClickListener: OnAdapterItemClickListener<*>? = null
    var orientation = 0
    var dialogAnimationRes = 0
    var dialogView: View? = null
    var onDismissListener: DialogInterface.OnDismissListener? = null
    var onKeyListener: DialogInterface.OnKeyListener? = null

    //////////////////////////////////////////Parcelable持久化//////////////////////////////////////////////////////
    constructor() {}
    protected constructor(`in`: Parcel) {
        layoutRes = `in`.readInt()
        height = `in`.readInt()
        width = `in`.readInt()
        dimAmount = `in`.readFloat()
        gravity = `in`.readInt()
        tag = `in`.readString()
        ids = `in`.createIntArray()
        isCancelableOutside = `in`.readByte().toInt() != 0
        orientation = `in`.readInt()
    }

    //内容描述接口,不用管
    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(layoutRes)
        dest.writeInt(height)
        dest.writeInt(width)
        dest.writeFloat(dimAmount)
        dest.writeInt(gravity)
        dest.writeString(tag)
        dest.writeIntArray(ids)
        dest.writeByte((if (isCancelableOutside) 1 else 0).toByte())
        dest.writeInt(orientation)
    }

    fun setAdapter(adapter: A) {
        this.adapter = adapter
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<XController<*>> =
            object : Parcelable.Creator<XController<*>> {
                override fun createFromParcel(`in`: Parcel): XController<*> {
                    return XController<XBaseAdapter<*>>(`in`)
                }

                override fun newArray(size: Int): Array<XController<*>?> {
                    return arrayOfNulls(size)
                }
            }
    }
}